### PR TITLE
[clang] Add tests for DRs about complete-class context

### PR DESCRIPTION
**Pull request by Endilll**
[P1787](https://wg21.link/p1787): The intent for CWG2335 (contra those of the older CWG1890, CWG1626, CWG1255, and CWG287) is supported by retaining the unrestricted forward lookup in complete-class contexts (despite current implementation behavior for non-templates).

Wording: The declaration set is the result of a single search in the scope of C for N from immediately after the class-specifier of C if P is in a complete-class context of C or from P otherwise. [Drafting note: The plan for CWG2335 is to describe forbidden dependency cycles among the complete-class contexts of a class. — end drafting note] ([class.member.lookup]/4)

Complete-class context is described in [class.mem.general] [p7](http://eel.is/c++draft/class#mem.general-7) and [p8](http://eel.is/c++draft/class#mem.general-8). In this patch I add tests only for CWG issues that fall under current definition of complete-class context, because I'm not sure how CWG1255 and CWG287 are going to work. That's why I skip over them, but mark CWG1308 as superseded by CWG1330.

### CODE REVIEW

**Comment by Endilll** [comment]
This PR has been migrated from https://reviews.llvm.org/D148433.
I made non-functional changes to `expected` directives to match the rest of the DR test suite.
I also moved CWG2335 test in a dedicated file in order to avoid issues with test isolation. Clang can stop instantiating templates if an error occurred earlier in a TU, which leads to less diagnostics emitted. Test for CWG2335 runs into this issue.

File: clang/test/CXX/drs/dr16xx.cpp [hunk 1]
```diff
@@ -42,6 +42,28 @@ namespace dr1611 { // dr1611: dup 1658
   C c;
 }
 
+namespace dr1626 { // dr1626: no open
+// FIXME: current consensus for CWG2335 is that the examples are well-formed.
```
Comment (cor3ntin) [inline clang/test/CXX/drs/dr16xx.cpp:46]: It is not clear from https://www.open-std.org/jtc1/sc22/wg21/docs/cwg_active.html#1626 what the intent for that is.

The minutes are very terse
```
- I agree with you, would be happy to mark as NAD
- Will mark as P1
```

**Review by cor3ntin — APPROVED** [review]
LGTM modulo comment

File: clang/test/CXX/drs/dr16xx.cpp (hunk 1 above)
Comment (Endilll) [reply clang/test/CXX/drs/dr16xx.cpp:46]: Intent expressed in P1787 in PR description seems rather clear to me, as well as its application to CWG1626, but I don't insist.

Comment (Endilll) [reply clang/test/CXX/drs/dr16xx.cpp]: We had an offline discussion, and concluded that this test should be removed.

**Comment by Endilll** [comment]
CI fails on formatting in unrelated files, so I'm going to ignore it:
```
t=1704988008448 + echo '*** Checking for trailing whitespace left in Clang source files ***'
t=1704988008448 *** Checking for trailing whitespace left in Clang source files ***
t=1704988008448 + grep -rnI '[[:blank:]]$' clang/lib clang/include clang/docs
t=1704988008595 clang/docs/HLSL/FunctionCalls.rst:147:Clang Implementation 
t=1704988008602 + echo '*** Trailing whitespace has been found in Clang source files as described above ***'
t=1704988008602 *** Trailing whitespace has been found in Clang source files as described above ***
t=1704988008602 + exit 1
t=1704988008603 ^^^ +++
t=1704988008603 Error: The command exited with status 1
t=1704988008603 ^^^ +++
t=1704988008603 user command error: exit status 1
```
https://buildkite.com/llvm-project/github-pull-requests/builds/28418

**Review by shafik — COMMENTED** [review]
LGTM

**Pull request merged**